hanish/darksky.py

```diff
@@ -0,0 +1,192 @@
+# hanish.darksky
+# Handlers and functionality for making Dark Sky API requests.
+#
+# Author:   Benjamin Bengfort <benjamin@bengfort.com>
+# Created:  Mon May 08 12:04:11 2017 -0400
+#
+# Copyright (C) 2016 Bengfort.com
+# For license information, see LICENSE.txt
+#
+# ID: cdarksky.py [] benjamin@bengfort.com $
+
+"""
+Handlers and functionality for making Dark Sky API requests.
+"""
+
+##########################################################################
+## Imports and Module Vars
+##########################################################################
+
+import requests
+
+try:
+    # Python 3
+    from urllib.parse import urljoin
+except ImportError:
+    # Python 2
+    from urlparse import urljoin
+
+
+from datetime import date, datetime
+from expiringdict import ExpiringDict
+from .exceptions import HanishValueError
+from .exceptions import DarkSkyException
+
+
+DARKSKY_API_URL  = "https://api.darksky.net"
+API_CALLS_HEADER = "X-Forecast-API-Calls"
+
+
+##########################################################################
+## DarkSky API
+##########################################################################
+
+class DarkSky(object):
```

Comment:
I think both answers were helpful. I used to consider caching on the client-side an anti-pattern, but I think for an external service like `darksky` it makes total sense. Indeed, we straight up copy data into our database, matching the more persistent caching option you describe. Once it's in our db, caching it with something like `redis`is natural.

My question was somewhat leading because I'm a novice at `python`, but I was wondering if this was a good use case for decorators. The annotative kind would probably not be appropriate, but would it be idiomatic in python to wrap and API client in a caching decorator.

The act of `put` and `get` could probably have multiple 🦆 -typed implementations. I'm not sure how tractable that is, but it would keep the concern of caching out of the implementation entirely. The configurable cache as part of the the client building make sense too.